src/sentry/sentry_metrics/indexer/postgres_v2.py

```diff
@@ -0,0 +1,260 @@
+        For a given key, add the id to the results.
+        """
+        org_id, string = key_pair
+        try:
+            self.results[org_id][string] = result
+        except KeyError:
+            self.results[org_id] = {string: result}
+
+    def get_unmapped_keys(self):
+        """
+        Return a new KeyCollection for any keys that don't have corresponding
+        ids in results.
+        """
+        unmapped_org_strings: MutableMapping[int, Set[str]] = defaultdict(set)
+        for org_id, strings in self.mapping.items():
+            for string in strings:
+                try:
+                    self.results[org_id][string]
+                except KeyError:
+                    unmapped_org_strings[org_id].add(string)
+
+        return KeyCollection(unmapped_org_strings)
+
+    def get_mapped_results(self) -> MutableMapping[int, MutableMapping[str, int]]:
+        """
+        Return the results.
+        """
+        return self.results
+
+    def get_mapped_key_strings_to_ints(self) -> MutableMapping[str, int]:
+        """
+        Return the results, but formatted as the following:
+            {
+                "1:a": 10,
+                "1:b": 11,
+                "1:c", 12,
+                "2:e": 13
+            }
+        This is for when we use indexer_cache.set_many()
+        """
+        cache_key_results: MutableMapping[str, int] = {}
+        for org_id, result_dict in self.results.items():
+            for string, id in result_dict.items():
+                key = f"{org_id}:{string}"
+                cache_key_results[key] = id
+
+        return cache_key_results
+
+
+class PGStringIndexerV2(Service):
+    """
+    Provides integer IDs for metric names, tag keys and tag values
+    and the corresponding reverse lookup.
+    """
+
+    __all__ = ("record", "resolve", "reverse_resolve", "bulk_record")
+
+    def _get_db_records(self, db_keys: KeyCollection) -> Any:
+        conditions = []
+        for pair in db_keys.as_tuples():
+            organization_id, string = pair
+            conditions.append(Q(organization_id=int(organization_id), string=string))
+
+        query_statement = reduce(or_, conditions)
+
+        return StringIndexerTable.objects.filter(query_statement)
+
+    def bulk_record(
+        self, org_strings: MutableMapping[int, Set[str]]
+    ) -> MutableMapping[int, MutableMapping[str, int]]:
+        """
+        TODO(meredith): add doc string
+        """
+        mapped_results: MutableMapping[int, MutableMapping[str, int]] = {}
+
+        cache_keys = KeyCollection(org_strings)
+
+        # Step 1: Lookup ids in the cache and add them to the
+        # cache KeyCollection results.
+        cache_hits = 0
+        cache_misses = 0
+        cache_key_strs = cache_keys.as_strings()
+        cache_results = indexer_cache.get_many(cache_key_strs)
+        for cache_key in cache_key_strs:
+            result = cache_results.get(cache_key)
+            org_id, string = cache_key.split(":")
```

Comment:
This responsibility should go to `KeyCollection`. IF that is the class that knows how to pack `org` and `string` into a single string it should also be the one that knows how to unpack them. Splitting these two tasks opens the door in introducing bugs because the two have to always be updated together (thus part of the same responsibility). It would be much harder to miss one if packing and unpacking were in the same class. 

Specifically I would add a static method to `KeyCollection` like :
```
def from_strings(cls, Sequence[str]) -> KeyCollection:
```
That is the opposite of `as_strings`

Then you can simply build the collection as 
```
hist = [k for k, v in cache_results.items() if v is not None]
KeyCollection::from_strings(hits)
```

Then cache hit is `len(hits)` while misses is `len(results) - len(hits)`